### ISSUE TITLE
squashfs image not working

### ISSUE THREAD

**Issue by hufnagel** [apptainer/singularity#957]
## Version of Singularity:

Package singularity.x86_64 0:2.3.1-0.1.el6

bash-4.1$ singularity --version
2.3.1-dist

### Expected behavior

I want to mount a squashfs image. This should work according to #192 

### Actual behavior

singularity doesn't recognize the image as valid.

bash-4.1$ file /storage/local/data1/cvmfs/rhel6.squashfs
/storage/local/data1/cvmfs/rhel6.squashfs: Squashfs filesystem, little endian, version 4.0, 363704024 bytes, 31213 inodes, blocksize: 131072 bytes, created: Sat Sep 23 16:52:26 2017

bash-4.1$ singularity -v shell /storage/local/data1/cvmfs/rhel6.squashfs
Increasing verbosity level (2)
Singularity version: 2.3.1-dist
Exec'ing: /usr/libexec/singularity/cli/shell.exec
Evaluating args: '/storage/local/data1/cvmfs/rhel6.squashfs'
VERBOSE: Set messagelevel to: 2
VERBOSE: Initialize configuration file: /etc/singularity/singularity.conf
VERBOSE: Initializing Singularity Registry
VERBOSE: Invoking the user namespace
VERBOSE: Not virtualizing USER namespace: running as SUID
VERBOSE: Using session directory: /var/singularity/mnt/session
VERBOSE: Checking what kind of image we are mounting
VERBOSE: File is not a valid Singularity image
ERROR  : Unknown image format/type.
ABORT  : Retval = 255

### Steps to reproduce behavior

The rhel6.squashfs image was created with

mksquashfs /cvmfs/singularity.opensciencegrid.org/bbockelm/cms\:rhel6/ rhel6.squashfs

**Comment by vsoch** [comment]
I don't think a version prior to 2.4 would support shell (the issue you linked to mentions export and import for example) you would want to use 2.4 (current development branch or later) to shell (and I see you are using 2.3.1-dist) @gmkurtzer you probably can comment on this.

**Comment by hufnagel** [comment]
The issue I linked to is about shell, there is even a working example.

**Comment by vsoch** [comment]
The working example is from September 10, which coincides with 2.3.2. 2.3.1 is from June 26, so I don't think squash stuffs had been done yet.

**Comment by vsoch** [comment]
There should be a `src/lib/rootfs/Makefile` with squashfs defined, it's not there in the 2.3.1 dist:

https://github.com/singularityware/singularity/tree/e214d4ebf0a1274b1c63b095fd55ae61c7e92947/src/lib

**Comment by hufnagel** [comment]
The working example is from September 10, 2016 :-)

**Comment by vsoch** [comment]
holy damn, you're right! Still, I'm looking at the code from the branch that @gmkurtzer referenced:

https://github.com/singularityware/singularity/commit/41fa9875e824e25056143dedcd90c3fc6abb199c#diff-fb5f08acf4ed45170f85a442ac13ee64

and it's completely different from what we had a 2.3.1. It looks like squashfs is now located here:

https://github.com/singularityware/singularity/blob/e214d4ebf0a1274b1c63b095fd55ae61c7e92947/src/lib/image/mount/squashfs/squashfs.c

But the relevant code should be here, where we try to mount:

https://github.com/singularityware/singularity/blob/e214d4ebf0a1274b1c63b095fd55ae61c7e92947/src/lib/image/mount/mount.c

But it looks like the image type found doesn't trigger any of the known types and it returns to image.c and prints the error message. The defined types are:

```
#define SINGULARITY 1
#define SQUASHFS    2
#define DIRECTORY   3
``` 

(in image.h). Can you try just peeping at the image header and looking at the number? There must be some issue that it doesn't parse 2 and thus detect the image.

**Comment by vsoch** [comment]
This might be a relevant issue - something to do with the offset: https://github.com/singularityware/singularity/issues/817

**Comment by hufnagel** [comment]
Did the fix for #817 make it into 2.3.2 ?

**Comment by vsoch** [comment]
Looks like the fix was:
```
    if ( image->type == SQUASHFS ) {
        singularity_message(DEBUG, "File is a SquashFS image, returning zero offset\n");
          return(0);
      }
```
from: https://github.com/singularityware/singularity/pull/820/files

and this isn't done in 2.3.1:

https://github.com/singularityware/singularity/blob/e214d4ebf0a1274b1c63b095fd55ae61c7e92947/src/lib/image/offset/offset.c

because it looks like @gmkurtzer might have lost it when re-writing the offset.c file, unless it's somewhere else?

https://github.com/singularityware/singularity/commit/07992e1c496826ce7feda116036f54ac9f22e769#diff-492809f2e8298e83cecb61c17905409c

Anyway, I think we've found the issue, and we just need some clarification from @gmkurtzer. Thanks for finding this bug @hufnagel ! Apologies on our side for fixing and promptly unfixing it, ha.

**Comment by gmkurtzer** [comment]
Heya guys,... this should be working now at least in the development branch which will shortly be released as 2.4. I'm gonna close the issue, and apologize for breaking it (again)! Ooopsie!
```
$ file foo.squashfs 
foo.squashfs: Squashfs filesystem, little endian, version 4.0, 46148889 bytes, 5478 inodes, blocksize: 131072 bytes, created: Wed Oct  4 09:43:04 2017
$ singularity exec foo.squashfs uname -a
Linux gmkdev2 4.2.0-27-generic #32~14.04.1-Ubuntu SMP Fri Jan 22 15:32:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

If in fact it is not fixed, and I'm bonkers, please re-comment and we can reopen. Thanks!
